File: Algorithm/c_or_cpp/charpter16_slide/level2/min_sub_array_len.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int minSubArrayLen(int target, int* nums, int numsSize) {
    int left = 0, right = 0, sum = 0, min = INT_MAX;
    while (right < numsSize) {
        sum += nums[right++];
        while (sum >= target) {
            min = std::min(min, right - left);
            sum -= nums[left++];
        }
    }
    return min == INT_MAX ? 0 : min;
}

int main() {
    int target = 10;
    int nums[] = {1, 2, 3, 4, 5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int result = minSubArrayLen(target, nums, numsSize);
    std::cout << "Length of longest substring with at most " << target << " distinct characters: " << result << std::endl;
    return 0;
}

File: Algorithm/Java/src/main/java/com/yugutou/charpter16_slide/level2/MinSubArrayLen.java
package com.yugutou.charpter16_slide.level2;

public class MinSubArrayLen {
    public static void main(String[] args) {
        int[] nums = {2, 3, 1, 2, 4, 3};
        System.out.println(minSubArrayLen(7, nums));
    }

    public static int minSubArrayLen(int target, int[] nums) {
        int left = 0, right = 0, sum = 0, min = Integer.MAX_VALUE;
        while (right < nums.length) {
            sum += nums[right++];
            while (sum >= target) {
                min = Math.min(min, right - left);
                sum -= nums[left++];
            }
        }
        return min == Integer.MAX_VALUE ? 0 : min;
    }
}
